pkg/patch/announce.go

```diff
@@ -0,0 +1,283 @@
+	SetRecipients(recipients ...string) error
+	Send(content, subject string) error
+}
+
+//counterfeiter:generate -o internal/internalfakes/fake_release_noter.go . ReleaseNoter
+type ReleaseNoter interface {
+	GetMarkdown() (string, error)
+}
+
+//counterfeiter:generate -o internal/internalfakes/fake_formatter.go . Formatter
+type Formatter interface {
+	MarkdownToHTML(markdown string, title string) (string, error)
+}
+
+//counterfeiter:generate -o internal/internalfakes/fake_workspace.go . Workspace
+type Workspace interface {
+	Status() (map[string]string, error)
+}
+
+// The date/time layouts to parse and format `time.Time`s
+// Reference date is 'Mon Jan 2 15:04:05 MST 2006'
+const (
+	dateLayoutISO8601    = "2006-01-02"
+	dateLayoutDayISO8601 = "Monday, 2006-01-02"
+)
+
+func dateFormatHuman(t time.Time) string {
+	return t.Format(dateLayoutDayISO8601)
+}
+
+func loadTemplate(tmplString string) (*template.Template, error) {
+	funcs := template.FuncMap{
+		"dateFormatHuman": dateFormatHuman,
+		"code":            func(s string) string { return "`" + s + "`" },
+		"codeBlock":       func(s string) string { return "```\n" + s + "\n```" },
+		"link":            func(n, t string) string { return "[" + n + "](" + t + ")" },
+	}
+	return template.New("main").Funcs(funcs).Parse(tmplString)
+}
+
+func (a *Announcer) getMailHead(version string, freezeDate, cutDate time.Time) (string, error) {
+	tmpl, err := loadTemplate(MailHeadMarkdown)
+	if err != nil {
+		return "", err
+	}
+
+	templated := &bytes.Buffer{}
+	templateData := struct {
+		DateFreeze                 time.Time
+		DateCut                    time.Time
+		Version                    string
+		ReleaseManagerName         string
+		ReleaseManagerEmail        string
+		ReleaseManagerTag          string
+		ReleaseManagerSlackChannel string
+	}{
+		DateFreeze:                 freezeDate,
+		DateCut:                    cutDate,
+		Version:                    version,
+		ReleaseManagerName:         ReleaseManagerName,
+		ReleaseManagerTag:          ReleaseManagerTag,
+		ReleaseManagerEmail:        ReleaseManagerEmail,
+		ReleaseManagerSlackChannel: ReleaseManagerSlackChannel,
+	}
+
+	err = tmpl.Execute(templated, templateData)
+	if err != nil {
+		return "", err
+	}
+
+	return templated.String(), nil
+}
+
+func (a *Announcer) getUpcomingVer() (string, error) {
+	if a.Workspace == nil {
+		w := &internal.Workspace{
+			K8sRepoPath: a.Opts.K8sRepoPath,
+		}
+		w.SetLogger(a.Logger(), "workspace")
+		a.Workspace = w
+		a.Logger().Debug("new workspace created")
+	}
+
+	status, err := a.Workspace.Status()
+	if err != nil {
+		return "", err
+	}
+
+	// v1.18.0-alpha.2.121+e92a7cfd2a82cd-dirty
+	gitVersion, ok := status["gitVersion"]
+	if !ok {
+		return "", fmt.Errorf("workspace status has no field 'gitVersion': %q", status)
+	}
+
+	i := strings.IndexRune(gitVersion, '-')
+	if i < 0 {
+		return "", fmt.Errorf("cannot extract upcoming version from gitVersion %q", gitVersion)
```

Comment:
make consistent with the above:
> `fmt.Errorf("workspace status has no field 'gitVersion': %q", status)`

Reply:
I think these are two distinct error cases:
1. we cannot even find the thing we are looking for, ie. `gitVersion`
1. the thing we are looking for is there, but it is not in the shape or form we expected it to be, ie. currently in this case that means it does not include a `-`

Reply:
one seems to have a quoted `gitVersion`

Reply:
oh ... got it! Will do this in a follow up!